### PR TITLE
[ds-fusion]Add pattern for offset as a function of loop iteration (ds fusion)

### PR DESCRIPTION
**Pull request by shraiysh**
Improving the pattern being recognized in dynamic slice fusion, to allow offset as a function of loop induction variable. This offset will later be calculated on the host at runtime (the host will keep track of the induction variable).

### CODE REVIEW

File: xla/service/gpu/transforms/dynamic_slice_fusion_rewriter.cc
Comment (bchetioui) [inline xla/service/gpu/transforms/dynamic_slice_fusion_rewriter.cc:205]: Instead of creating two new modules, we could just walk backwards from `update` and `value` all the way to the `while_body`'s tuple parameter? If we got there by any other path than through the right `GTE`, then we return `false`. Similarly if one of the instructions is a `replica-id` or a `partition-id`.

That'd be cheaper, and not very complex, I think. WDYT?

File: xla/service/gpu/transforms/dynamic_slice_fusion_rewriter.cc
Comment (bchetioui) [inline xla/service/gpu/transforms/dynamic_slice_fusion_rewriter.cc:226]: Can you expand on why? My assumption is that this is because it inherently introduces additional, implicit parameters to the computation.

File: xla/service/gpu/transforms/dynamic_slice_fusion_rewriter_test.cc
Comment (bchetioui) [inline xla/service/gpu/transforms/dynamic_slice_fusion_rewriter_test.cc:2146]: Can you specify what is the behaviour that you expect to test here in the test name?

File: xla/service/gpu/transforms/dynamic_slice_fusion_rewriter_test.cc
Comment (bchetioui) [inline xla/service/gpu/transforms/dynamic_slice_fusion_rewriter_test.cc:2148]: Seems like a lot of HLO---do we need all of it to exhibit the behaviour you intend to test?

**Review by bchetioui — CHANGES_REQUESTED** [review]

**Comment by bchetioui** [comment]
Additional request: could you expand your thinking a little more in the commit message? We'd like the commit message to make sense in a vacuum, and without having full context on what the file is doing :)

File: xla/service/gpu/transforms/dynamic_slice_fusion_rewriter.cc
Comment (shraiysh) [reply xla/service/gpu/transforms/dynamic_slice_fusion_rewriter.cc]: Yes, thanks for pointing it out.

File: xla/service/gpu/transforms/dynamic_slice_fusion_rewriter.cc
Comment (shraiysh) [reply xla/service/gpu/transforms/dynamic_slice_fusion_rewriter.cc]: Yes, thanks for the suggestion. Implemented it that way.

File: xla/service/gpu/transforms/dynamic_slice_fusion_rewriter.cc
Comment (shraiysh) [reply xla/service/gpu/transforms/dynamic_slice_fusion_rewriter.cc]: Yes, you are correct. The host needs to know this (shouldn't be hard, but trying to support that will be the next step, after this offset work gets done.)

File: xla/service/gpu/transforms/dynamic_slice_fusion_rewriter_test.cc
Comment (shraiysh) [reply xla/service/gpu/transforms/dynamic_slice_fusion_rewriter_test.cc]: Done. Thanks.

File: xla/service/gpu/transforms/dynamic_slice_fusion_rewriter_test.cc
Comment (shraiysh) [reply xla/service/gpu/transforms/dynamic_slice_fusion_rewriter_test.cc:2148]: Unfortunately, yes. I need the while loop, the collective, and a simple (yet random) seq of operations for making sure it is being fused as a general pattern.

**Comment by shraiysh** [comment]
> Additional request: could you expand your thinking a little more in the commit message? We'd like the commit message to make sense in a vacuum, and without having full context on what the file is doing :)

I changed it. Let me know if you'd like me to add something.

File: xla/service/gpu/transforms/dynamic_slice_fusion_rewriter.cc
Comment (bchetioui) [inline xla/service/gpu/transforms/dynamic_slice_fusion_rewriter.cc:165]: With the new implementation, checking for `replica-id` and `partition-id` specifically seems unnecessary for the current PR.

The condition is essentially:

```cpp
if (consumer.operand_count() == 0) {
  return false;
}
```

How about we just simplify the API like this? We can implement special support later, if it's necessary.

File: xla/service/gpu/transforms/dynamic_slice_fusion_rewriter.cc
Comment (bchetioui) [inline xla/service/gpu/transforms/dynamic_slice_fusion_rewriter.cc:166]: As specified in a previous PR, we'd like to have braces around all controlled statements (even single-line). Please fix this here and below.

File: xla/service/gpu/transforms/dynamic_slice_fusion_rewriter.cc
Comment (bchetioui) [inline xla/service/gpu/transforms/dynamic_slice_fusion_rewriter.cc:253]: Let's use `absl::c_all_of` here too (it is inconsistent to use two different methods to do this in the same file).

File: xla/service/gpu/transforms/dynamic_slice_fusion_rewriter_test.cc
Comment (bchetioui) [inline xla/service/gpu/transforms/dynamic_slice_fusion_rewriter_test.cc:2146]: nit: Indvar -> InductionVariable

**Review by bchetioui — CHANGES_REQUESTED** [review]
Looks much better @shraiysh! Just a few more minor requests.

Also another one: please don't squash when iterating on review comments---otherwise it is hard to see what changed :)

File: xla/service/gpu/transforms/dynamic_slice_fusion_rewriter.cc
Comment (shraiysh) [reply xla/service/gpu/transforms/dynamic_slice_fusion_rewriter.cc]: Done

File: xla/service/gpu/transforms/dynamic_slice_fusion_rewriter.cc
Comment (shraiysh) [reply xla/service/gpu/transforms/dynamic_slice_fusion_rewriter.cc]: Sorry about that. Missed it here. I think it is all covered now, but please let me know if you spot anywhere I missed it. (I will check if we can we get clang-format/clang-tidy to check for this?)

File: xla/service/gpu/transforms/dynamic_slice_fusion_rewriter.cc
Comment (shraiysh) [reply xla/service/gpu/transforms/dynamic_slice_fusion_rewriter.cc]: Done. Thanks.

**Comment by shraiysh** [comment]
Requesting review again, addressed comments.

**Review by bchetioui — CHANGES_REQUESTED** [review]
Thank you @shraiysh, this looks great! I left you one last small comment, please ping me once you've addressed it and I will approve.

(Unfortunately can't approve without resolving everything first, as this will trigger a merge request automatically.)

File: xla/service/gpu/transforms/dynamic_slice_fusion_rewriter.cc
Comment (shraiysh) [reply xla/service/gpu/transforms/dynamic_slice_fusion_rewriter.cc]: Done.

**Review by bchetioui — APPROVED** [review]
Thanks!

**Pull request closed without merging**
